samples/WinForms/Dx12Panel.cs

```diff
@@ -0,0 +1,761 @@
+                            D3D12_HEAP_FLAG_NONE,
+                            &bufferDesc,
+                            D3D12_RESOURCE_STATE_GENERIC_READ,
+                            pOptimizedClearValue: null,
+                            &iid,
+                            (void**)vertexBuffer
+                        ));
+                    }
+
+                    // Copy the triangle data to the vertex buffer.
+                    var readRange = new D3D12_RANGE();
+
+                    byte* pVertexDataBegin;
+                    ThrowIfFailed(nameof(ID3D12Resource._Map), _vertexBuffer->Map(Subresource: 0, &readRange, (void**)&pVertexDataBegin));
+                    Unsafe.CopyBlock(pVertexDataBegin, triangleVertices, vertexBufferSize);
+                    _vertexBuffer->Unmap(0, null);
+
+                    // Initialize the vertex buffer view.
+                    _vertexBufferView.BufferLocation = _vertexBuffer->GetGPUVirtualAddress();
+                    _vertexBufferView.StrideInBytes = (uint)sizeof(Vertex);
+                    _vertexBufferView.SizeInBytes = vertexBufferSize;
+                }
+
+                // Create synchronization objects and wait until assets have been uploaded to the GPU.
+                {
+                    fixed (ID3D12Fence** fence = &_fence)
+                    {
+                        iid = IID_ID3D12Fence;
+                        ThrowIfFailed(nameof(ID3D12Device.CreateFence), _device->CreateFence(0, D3D12_FENCE_FLAG_NONE, &iid, (void**)fence));
+                        _fenceValue = 1;
+                    }
+
+                    // Create an event handle to use for frame synchronization.
+                    _fenceEvent = CreateEventW(lpEventAttributes: null, bManualReset: FALSE, bInitialState: FALSE, lpName: null);
+                    if (_fenceEvent == null)
+                    {
+                        var hr = Marshal.GetHRForLastWin32Error();
+                        Marshal.ThrowExceptionForHR(hr);
+                    }
+
+                    // Wait for the command list to execute; we are reusing the same command
+                    // list in our main loop but for now, we just want to wait for setup to
+                    // complete before continuing.
+                    WaitForPreviousFrame();
+                }
+            }
+            finally
+            {
+                if (signature != null)
+                {
+                    signature->Release();
+                }
+
+                if (error != null)
+                {
+                    error->Release();
+                }
+
+                if (vertexShader != null)
+                {
+                    vertexShader->Release();
+                }
+
+                if (pixelShader != null)
+                {
+                    pixelShader->Release();
+                }
+            }
+        }
+
+        private void OnResize(object? sender, EventArgs e)
```

Comment:
This should override `OnSizeChanged` (or maybe `OnClientSizeChanged`, since that is the area that we really care about).

Reply:
I switched it to overriding OnClientSizeChanged(). It triggers calling OnPaint automatically, so no extra call to that needed. Seems indeed better using this inheritance mechanism than adding an event handler.